src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs

```diff
@@ -0,0 +1,96 @@
+
+        /// <summary>
+        /// The test tfm to select.
+        /// </summary>
+        /// This is the TestTFM folder that we want to select test zips from. This needs to be on the path to the test zip.
+        [Required]
+        public string FilterToTestTFM { get; set; }
+
+        /// <summary>
+        /// The OSGroup to select.
+        /// </summary>
+        /// This is the OSGroup folder that we want to select test zips from. This does not require an exact match as some tests 
+        /// may or may not have an exact OSGroup folder match.
+        [Required]
+        public string FilterToOSGroup { get; set; }
+
+        /// <summary>
+        /// The test archives folder.
+        /// </summary>
+        /// This is the path under a given Platform/OSGroup/Configuration folder where the Test Archives are placed
+        [Required]
+        public string TestArchivesRoot { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems { get; set; }
+
+        private readonly List<ITaskItem> FilteredUploadItemsList = new List<ITaskItem>();
+
+        public override bool Execute()
+        {
+            string[] filesToSearch = Directory.GetFiles(TestOutputDirListsLocation);
+            
+            foreach (var filename in filesToSearch)
+            {
+                string selectedTest = string.Empty;
+                using (StreamReader sr = new StreamReader(File.OpenRead(filename)))
+                {
+                    string line;
+                    while ((line = sr.ReadLine()) != null)
+                    {
+                        // Each line is of the format "tests/AnyOS.AnyCPU.Debug/Example.Tests/netcoreapp1.0"
+                        // Each test zip is of the format "tests/AnyOS.AnyCPU.Debug/archive/tests/netcoreapp1.0/Example.Tests.zip"
+                        if (CalculateMatchValue(line) > CalculateMatchValue(selectedTest))
+                        {
+                            string parentDirectory = Path.GetDirectoryName(line);
+                            string testTfmFolderInArchive = Path.GetFileName(line);
+                            string testArchiveDir = Path.Combine(
+                                Path.GetDirectoryName(parentDirectory),
+                                TestArchivesRoot);
+                            string testZipName = Path.GetFileName(parentDirectory) + ".zip";
+                            selectedTest = Path.Combine(testArchiveDir, testTfmFolderInArchive, testZipName);
+                        }
+                    }
+                }
+                if (!string.IsNullOrEmpty(selectedTest))
+                {
+                    FilteredUploadItemsList.Add(new TaskItem(selectedTest));
+                }
+            }
+
+            FilteredUploadItems = FilteredUploadItemsList.ToArray();
+            return true;
+        }
+        
+        private int CalculateMatchValue(string testPathToMatch)
```

Comment:
This still isn't safe, the "Example.Test" might be "CheckUnixWindows.Tests", which is going to do really, really bad things all over the place.